Pdb 25 account creation (student flow)

heidage opened the pull request:
1) Added in JWT passing between frontend and backend
2) Added in account initialisation after clicking either Student / Tutor
3) Added in account creation depending on which account type user clicks on

heidage committed:
added in JWT passing between frontend & backend

heidage committed:
added in account initialise and creation for student and tutors

TsienJin commented on Bamboo/frontend/bamboo-frontend/app/routes/_setup+/welcome.tsx:
```diff
@@ -30,6 +54,11 @@ export default function SetupWelcomePage(){
+      auth.initilaise.accountType(
+        BACKEND,
+        selectState,
+        jwt
```
iirc http-only cookies will be sent to the backend if we set "auth:true" when making a request

TsienJin replied:
Set to be HTTP only "false" for debugging

TsienJin commented on Bamboo/backend/bamboo-backend/src/routes/auth/initialise/initialise.controller.ts:
```diff
@@ -0,0 +1,35 @@
+    ){
+        const decoded: JwtPayload = jwt.decode(body.jwt) as JwtPayload;
+        try {
+            const { accountJwt } = await this.initialiseService.initialiseAccount(
```
I'll combine the JWT logic to a middleware or interceptor so this code can be isolated

TsienJin commented on Bamboo/backend/bamboo-backend/src/routes/auth/initialise/initialise.service.ts:
```diff
@@ -0,0 +1,36 @@
+
+        // Generate a new JWT token with the updated account
+        const firstName = account.name.split(' ')[0]
+        const lastName = account.name.split(' ')[1]
```
dont use string.split methods when determining firstName and lastName. These should be separate attributes

TsienJin replied:
userService.initialiseAccount response type should match the response type of userService.createUser or .findOrCreate

TsienJin commented on Bamboo/backend/bamboo-backend/src/services/user/user.service.ts:
```diff
@@ -49,4 +50,41 @@ export class UserService {
+    )
+
+    let account = null
+    if (accountType === 'student') {
```
Use switch instead

TsienJin commented on Bamboo/backend/bamboo-backend/src/services/user/user.service.ts:
Reference the code on the latest staging -- do `npm run prisma` if u've yet to run that. The updated schema should allow for optional passwords. Setting password as `''`  will result in a hash for an empty string. This is a vulnerability as a user can use an empty string to log in.

heidage replied:
I was thinking of using the code from the lastest staging but the schema.prisma doesnt have `password String?` for `Tutor` and `Student`.The `name` as well is not `firstName` and `lastName`. I was thinking maybe we could use the User object for registration and login, and having the emil as secondary key hence using it which links us to individual `Student` or `Tutor` record.

TsienJin commented on Bamboo/backend/bamboo-backend/src/services/user/user.service.ts:
```diff
@@ -49,4 +50,41 @@ export class UserService {
+    if (accountType === 'tutor') {
+      account = await this.prismaService.prisma.tutor.create({
+        data: {
+          name: `${jwt.firstName} ${jwt.lastName}`,
```
use firstName and lastName instead

TsienJin commented on Bamboo/backend/bamboo-backend/src/services/user/user.service.ts:
```diff
@@ -49,4 +50,41 @@ export class UserService {
   
   }
+
+  async initialiseAccount(accountType: string, jwt: JwtPayload){
```
initialiseAccount should not contain logic to create the `Student`/`Tutor` record in the DB -- this should be done in a separate service; one for student (e.g. studentService) and one for tutor (e.g. tutorService)

heidage committed:
seperate services to create student and tutor

heidage committed:
added in switch statements for student and tutor type, added in new migrations

heidage committed:
added in Student and Tutor Service to InitialiseModule

TsienJin committed:
PDB-25 Created UserAccountType enum to replace strings

TsienJin committed:
PDB-25 Updated prisma schema

- Made emails an indexed field for User, Student and Tutor
- Removed password and name fields on Student and Tutors

TsienJin committed:
PDB-25 Created standalone JWT service to isolate JWT logic

TsienJin committed:
PDB-25 Added PandaJwtService to app module

TsienJin committed:
PDB-25 Updated JWT signature and account initialisation

- Refactored JWT generation to use PandaJwtService
- Updated JWT to account for user type and add user id
- Changed the control flow of account initialisation to be after the DB record has been created
